File: Python/Clone Graph.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        return self.cloneHelper(node, {})
    
    def cloneHelper(self, node, visited):
        if not node:
            return None
           
        retNode = UndirectedGraphNode(node.label)
        visited[node.label] = retNode
        
        for neighbor in node.neighbors:
            if not neighbor.label in visited.keys():
                neighbornode = self.cloneHelper( neighbor, visited)
                visited[neighbor.label] = neighbornode
            retNode.neighbors.append(visited[neighbor.label])
        return retNode
            
        

File: Java/src/Clone Graph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        return cloneHelper(node, new HashMap<Integer,UndirectedGraphNode>());
        
    }
    private UndirectedGraphNode cloneHelper(UndirectedGraphNode node, Map<Integer,UndirectedGraphNode> visited){
        if (node == null) return null;
        UndirectedGraphNode retNode = new UndirectedGraphNode(node.label);
        visited.put(node.label, retNode);
        for (UndirectedGraphNode neighbor: node.neighbors){
            if (visited.get(neighbor.label) == null){
                UndirectedGraphNode neighborNode = cloneHelper(neighbor, visited);
                visited.put(neighbor.label, neighborNode);
            }
            retNode.neighbors.add(visited.get(neighbor.label));
        }
        return retNode;
        
        
    }
}